File: src/jimin/week5/프렌즈4블록.py
def solution(m, n, board):
    for i in range(m):
        board[i] = list(board[i])

    rm = set()
    answer = 0
    while True:
        for i in range(m - 1):
            for j in range(n - 1):
                current = board[i][j]
                if current == 'X':
                    continue
                #사각형 모양으로 저장
                if current == board[i + 1][j] and current == board[i][j + 1] and current == board[i + 1][j + 1]:
                    rm.add((i, j))
                    rm.add((i + 1, j))
                    rm.add((i, j + 1))
                    rm.add((i + 1, j + 1))

        if rm:  # 4x4 지웠다
            answer += len(rm)
            for i, j in rm:
                board[i][j] = 'X'
            rm = set()
        else:
            return answer

        # 새로 시작하기 전 빈칸 채우기
        while True:
            moved = 0
            for i in range(m - 1):
                for j in range(n):
                    if board[i + 1][j] == 'X' and board[i][j] != 'X':
                        board[i + 1][j] = board[i][j]
                        board[i][j] = 'X'
                        moved += 1
            if moved == 0:
                break


m = 4
n = 5
board = ["CCBDE", "AAADE", "AAABF", "CCBBF"]

m = 6
n = 6
board = ["TTTANT", "RRFACC", "RRRFCC", "TRRRAA", "TTMMMF", "TMMTTJ"]
print(solution(m, n, board))

File: src/geonhyeong/week5/프렌즈4블록.swift
/*
*   프렌즈4블록.swift
*  Question Link: https://school.programmers.co.kr/learn/courses/30/lessons/17679
*  Primary idea:   완탐
*                   1. 
*
*  걸린 시간 : 
*  Time Complexity : O()
*  Space Complexity : O()
*  Runtime:  ms
*  Memory Usage:  MB
*/

func solution(_ m:Int, _ n:Int, _ board:[String]) -> Int {
  var _board = board.map {Array($0).map {String($0)}}
  
  // 사각형(2x2)이 되는지 확인
  func isSquare(_ row: Int, _ col: Int) -> Bool {
    guard case let leftTop = _board[row][col], leftTop != "" else { // 비어 있을 경우
      return false
    }
    
    return _board[row][col + 1] == leftTop && _board[row + 1][col] == leftTop && _board[row + 1][col + 1] == leftTop
  }
  
  // 검사가 끝날때까지 무한 반복
  while true {
    var posList = [(row: Int, col:Int)]() // 사각형(2x2)이 되는 (row, col)
    
    // 사각형(2x2)의 leftTop 찾기
    for row in 0..<m-1 {
      for col in 0..<n-1 {
        if isSquare(row, col) {
          posList.append((row, col))
        }
      }
    }
    
    if posList.count == 0 { break }
    
    // 사각형(2x2) 빈칸으로 변경
    for pos in posList {
      _board[pos.row][pos.col] = ""
      _board[pos.row][pos.col+1] = ""
      _board[pos.row+1][pos.col] = ""
      _board[pos.row+1][pos.col+1] = ""
    }
    
    // 빈공간 채우기
    for col in 0..<n {
      for row in 1..<m { // 첫번쨰 행은 할 필요없음
        if _board[row][col] == "" {
          for _row in (1...row).reversed() {
            _board[_row][col] = _board[_row-1][col]
            _board[_row-1][col] = ""
          }
        }
      }
    }
  }

  return _board.reduce(0, {$0 + $1.filter {$0 == ""}.count})
}

//print(solution(4, 5, ["CCBDE", "AAADE", "AAABF", "CCBBF"]  )) // 14
//print(solution(6, 6, ["TTTANT", "RRFACC", "RRRFCC", "TRRRAA", "TTMMMF", "TMMTTJ"])) // 15
//print(solution(10, 10, ["DDABBAABBA","AAAAAABBBA","DDACCBBBAA","DDABBBBBAA","AAABBABBBA","CCADDAABBB","CCADDAABBB","BBACCABBBA","BBAAABBBAA","DDABBBBAAA"])) // 86
//print(solution(8, 5, ["HGNHU", "CRSHV", "UKHVL", "MJHQB", "GSHOT", "MQMJJ", "AGJKK", "QULKK"]))// 8
